segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/filesystem/FSStorageTest.java

```diff
@@ -0,0 +1,314 @@
+        val concatHandle = storage.openWrite(concatName).join();
+        storage.write(concatHandle, 0, new ByteArrayInputStream(data), data.length, TIMEOUT).join();
+        storage.seal(concatHandle, TIMEOUT).join();
+        storage.concat(handle, si.getLength() + data.length, concatHandle.getSegmentName(), TIMEOUT).join();
+    }
+
+    private void verifyWriteOperationsFail(SegmentHandle handle, Storage storage) {
+        val si = storage.getStreamSegmentInfo(handle.getSegmentName(), TIMEOUT).join();
+        final byte[] data = "hello".getBytes();
+        AssertExtensions.assertThrows(
+                "Write was not fenced out.",
+                () -> storage.write(handle, si.getLength(), new ByteArrayInputStream(data), data.length, TIMEOUT),
+                ex -> ex instanceof StorageNotPrimaryException);
+
+        // Create a second segment and try to concat it into the primary one.
+        final String concatName = "concat";
+        storage.create(concatName, TIMEOUT).join();
+        val concatHandle = storage.openWrite(concatName).join();
+        storage.write(concatHandle, 0, new ByteArrayInputStream(data), data.length, TIMEOUT).join();
+        storage.seal(concatHandle, TIMEOUT).join();
+        AssertExtensions.assertThrows(
+                "Concat was not fenced out.",
+                () -> storage.concat(handle, si.getLength(), concatHandle.getSegmentName(), TIMEOUT),
+                ex -> ex instanceof StorageNotPrimaryException);
+        storage.delete(concatHandle, TIMEOUT).join();
+    }
+
+    private void verifyFinalWriteOperationsSucceed(SegmentHandle handle, Storage storage) {
+        storage.seal(handle, TIMEOUT).join();
+        storage.delete(handle, TIMEOUT).join();
+
+        boolean exists = storage.exists(handle.getSegmentName(), TIMEOUT).join();
+        Assert.assertFalse("Segment still exists after deletion.", exists);
+    }
+
+    private void verifyFinalWriteOperationsFail(SegmentHandle handle, Storage storage) {
+        AssertExtensions.assertThrows(
+                "Seal was allowed on fenced Storage.",
+                () -> storage.seal(handle, TIMEOUT),
+                ex -> ex instanceof StorageNotPrimaryException);
+
+        val si = storage.getStreamSegmentInfo(handle.getSegmentName(), TIMEOUT).join();
+        Assert.assertFalse("Segment was sealed after rejected call to seal.", si.isSealed());
+
+        AssertExtensions.assertThrows(
+                "Delete was allowed on fenced Storage.",
+                () -> storage.delete(handle, TIMEOUT),
+                ex -> ex instanceof StorageNotPrimaryException);
+        boolean exists = storage.exists(handle.getSegmentName(), TIMEOUT).join();
+        Assert.assertTrue("Segment was deleted after rejected call to delete.", exists);
+    }
+
+    /**
+     * Tests the write() method.
+     *
+     * @throws Exception if an unexpected error occurred.
+     */
+    @Override
+    @Test
+    public void testWrite() throws Exception {
```

Comment:
In this case, only the order of calls is same. The behavior is different. Of course it can still be merged to use some common code, but I could not find any natural way to do this. If you have any recommendation, will follow that.